File: app/src/main/java/com/iamkatrechko/clipboardmanager/domain/service/ClipboardService.java
package com.iamkatrechko.clipboardmanager.domain.service;

import android.app.Service;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.IBinder;
import android.util.Log;
import android.widget.Toast;

import com.iamkatrechko.clipboardmanager.data.model.SimpleClip;
import com.iamkatrechko.clipboardmanager.view.NotificationManager;
import com.iamkatrechko.clipboardmanager.R;
import com.iamkatrechko.clipboardmanager.data.repository.ClipboardRepository;
import com.iamkatrechko.clipboardmanager.domain.util.ClipUtils;
import com.iamkatrechko.clipboardmanager.domain.util.UtilPreferences;

/**
 * Сервис для прослушки буфера обмена
 * @author iamkatrechko
 *         Date: 07.11.2016
 */
public class ClipboardService extends Service {

    /** Команда отображения избранных заметок */
    public static final String ACTION_SHOW_ONLY_FAVORITE = "action_show_only_favorite";
    /** Команда копирования заметки в буфер */
    public static final String ACTION_COPY_TO_CLIPBOARD = "action_copy_to_clipboard";
    /** Тег для логирования */
    private static final String TAG = ClipboardService.class.getSimpleName();
    /** Идентификатор уведомления сервиса */
    private static final int NOTIFICATION_ID = 98431;
    /** Менеджер буфера обмена */
    private ClipboardManager clipBoard;
    /** Репозиторий с записями в базе данных */
    private ClipboardRepository repository;
    /** Слушатель новых записей в буфере обмена */
    private ClipboardManager.OnPrimaryClipChangedListener clipListener = new ClipboardManager.OnPrimaryClipChangedListener() {
        @Override
        public void onPrimaryClipChanged() {
            Log.d(TAG, "Отловлена новая запись в буфере обмена");
            SimpleClip clip = ClipUtils.getClip(ClipboardService.this);

            Context context = getApplicationContext();
            if (clip.getLabel().equals(ClipUtils.CLIP_LABEL) || clip.getLabel().equals(ClipUtils.CLIP_LABEL_ACCESSIBILITY)) {
                Log.d(TAG, "Отмена: копирование из приложения");
                Toast.makeText(context, "Отмена: копирование из приложения", Toast.LENGTH_SHORT).show();
            } else {
                if (clip.getText().length() == 0) {
                    Toast.makeText(context, R.string.empty_record, Toast.LENGTH_SHORT).show();
                    // Обновляем уведомление
                    ClipboardService.startMyService(context);
                    return;
                }
                if (repository.alreadyExists(context, clip.getText())) {
                    //TODO Сделать настройку "Если уже существует: ничего не делать || изменить дату на новую
                    Toast.makeText(context, R.string.record_already_exists, Toast.LENGTH_SHORT).show();
                    // Обновляем уведомление
                    ClipboardService.startMyService(context);
                    return;
                }
                addNewClip(clip.getText());
            }
            // Обновляем уведомление
            ClipboardService.startMyService(context);
        }
    };

    /**
     * Запускает сервис
     * @param context контекст
     */
    public static void startMyService(Context context) {
        context.startService(new Intent(context, ClipboardService.class));
    }

    @Override
    public void onCreate() {
        Log.d(TAG, "onCreate");
        repository = ClipboardRepository.Companion.getInstance();
        clipBoard = (ClipboardManager) getSystemService(CLIPBOARD_SERVICE);
        clipBoard.addPrimaryClipChangedListener(clipListener);
        // TODO Отображать уведомление при запуске
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Log.d(TAG, "onStartCommand");

        String action = intent.getAction();
        if (action != null) {
            if (action.equals(ACTION_SHOW_ONLY_FAVORITE)) {
                boolean show = UtilPreferences.isShowOnlyFavoriteInNotification(this);
                UtilPreferences.setShowOnlyFavoriteInNotification(this, !show);
                startMyService(this);
            }
            if (action.equals(ACTION_COPY_TO_CLIPBOARD)) {
                long id = intent.getLongExtra("id", -1);
                String text = repository.getClip(this, id);
                if (text != null) {
                    ClipUtils.copyToClipboard(this, text);
                }
                startMyService(this);
            }
            return Service.START_NOT_STICKY;
        }
        boolean displayNotification = UtilPreferences.getDisplayNotification(this);

        startForeground(NOTIFICATION_ID, new NotificationManager().getNotification(this));

        if (!displayNotification) {
            startService(HideNotificationService.newIntent(this, NOTIFICATION_ID));
        }/*else{
            startForeground(98431, notification);
        }*/

        return Service.START_NOT_STICKY;
    }

    /**
     * Добавление новой записи в базу данных
     * @param text текст записи
     */
    private void addNewClip(String text) {
        Uri newClipUri = repository.addClip(this, text);
        if (newClipUri != null) {
            startService(CancelViewService.newIntent(this, newClipUri.getLastPathSegment()));
        }
    }

    @Override
    public IBinder onBind(Intent intent) {
        throw new UnsupportedOperationException("Не реализовано");
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        Log.i(TAG, "onDestroy");
        clipBoard.removePrimaryClipChangedListener(clipListener);
    }

    @Override
    public void onTaskRemoved(Intent rootIntent) {
        Log.i(TAG, "onTaskRemoved");
    }
}

File: app/src/main/java/com/iamkatrechko/clipboardmanager/domain/service/ClipboardService.kt
package com.iamkatrechko.clipboardmanager.domain.service

import android.app.Service
import android.content.Context
import android.content.Intent
import android.util.Log
import com.iamkatrechko.clipboardmanager.App
import com.iamkatrechko.clipboardmanager.data.model.SimpleClip
import com.iamkatrechko.clipboardmanager.data.repository.ClipboardRepository
import com.iamkatrechko.clipboardmanager.domain.repository.SystemsClipsWatcher
import com.iamkatrechko.clipboardmanager.domain.request.InsertClipRequest
import com.iamkatrechko.clipboardmanager.domain.util.ClipUtils
import com.iamkatrechko.clipboardmanager.domain.util.PrefsManager
import com.iamkatrechko.clipboardmanager.domain.util.SettingsValues
import com.iamkatrechko.clipboardmanager.view.NotificationManager
import com.iamkatrechko.clipboardmanager.view.extension.TAG
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.addTo

/**
 * Сервис для прослушивания буфера обмена
 * @author iamkatrechko
 *         Date: 07.11.2016
 */
class ClipboardService : Service() {

    /** Репозиторий с записями в базе данных */
    private val repository = ClipboardRepository.getInstance()
    /** Хранилище настроек программы */
    private val settings = SettingsValues.getInstance()
    /** Общий список rx-подписчиков */
    private val disposables = CompositeDisposable()

    override fun onCreate() {
        Log.d(TAG, "onCreate")
        val clipManager = SystemsClipsWatcher(App.clipManager, this.applicationContext)
        clipManager.observe()
                .subscribe(::prepareClip)
                .addTo(disposables)
        // TODO Отображать уведомление при запуске
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        Log.d(TAG, "onStartCommand")
        when (intent?.action) {
            ACTION_SHOW_ONLY_FAVORITE -> showOnlyFavorite()
            ACTION_COPY_TO_CLIPBOARD -> copyToClipboard(intent.getLongExtra("id", -1))
            else -> startForegroundService()
        }
        return START_NOT_STICKY
    }

    override fun onBind(intent: Intent) =
            null

    override fun onDestroy() {
        super.onDestroy()
        Log.d(TAG, "onDestroy")
        disposables.dispose()
    }

    override fun onTaskRemoved(rootIntent: Intent) {
        Log.d(TAG, "onTaskRemoved")
    }

    /** Обработчик найденных записей буфера */
    private fun prepareClip(clip: SimpleClip) {
        Log.d(TAG, "Обнаружена новая запись в буфере обмена")
        val (text, label) = clip

        if (label == ClipUtils.CLIP_LABEL || label == ClipUtils.CLIP_LABEL_ACCESSIBILITY) {
            Log.d(TAG, "Отмена: копирование из приложения")
        } else if (text.isEmpty()) {
            Log.d(TAG, "Отмена: текст записи пуст")
        } else if (repository.alreadyExists(text)) {
            Log.d(TAG, "Отмена: текущая запись уже существует")
            //TODO Сделать настройку "Если уже существует: ничего не делать || изменить дату на новую
        } else {
            addClipAndShowCancel(text)
        }
        // Обновляем уведомление
        startMyService(this)
    }

    /** Запускает фоновый сервис */
    private fun startForegroundService() {
        startForeground(NOTIFICATION_ID, NotificationManager().getNotification(this))
        if (!settings.notificationShow) {
            startService(HideNotificationService.newIntent(this, NOTIFICATION_ID))
        }
    }

    /** Включает режим отображения избранных записей */
    private fun showOnlyFavorite() {
        val prefs = PrefsManager.getInstance()
        prefs.isShowOnlyFavoriteInNotification = !prefs.isShowOnlyFavoriteInNotification
        startMyService(this)
    }

    /** Копирует запись с указанным [id] в буфер обмена */
    private fun copyToClipboard(id: Long) {
        repository.getClip(id)?.let {
            App.clipManager.toClipboard(it.text)
        }
        startMyService(this)
    }

    /** Добавление новой записи с текстом [content] в базу данных */
    private fun addClipAndShowCancel(content: String) {
        val clipId = repository.insertClip(InsertClipRequest.withContent(content))
        if (clipId != null) {
            startService(CancelViewService.newIntent(this, clipId.toString()))
        }
    }

    companion object {

        /** Команда отображения избранных заметок */
        const val ACTION_SHOW_ONLY_FAVORITE = "ACTION_SHOW_ONLY_FAVORITE"
        /** Команда копирования заметки в буфер */
        const val ACTION_COPY_TO_CLIPBOARD = "ACTION_COPY_TO_CLIPBOARD"
        /** Идентификатор уведомления сервиса */
        private const val NOTIFICATION_ID = 98431

        /** Запускает сервис */
        fun startMyService(context: Context) {
            context.startService(Intent(context, ClipboardService::class.java))
        }
    }
}
